### ISSUE TITLE
Detect Fediverse URLs by domain

### ISSUE THREAD

**Issue by Minoru** [tuskyapp/Tusky#2734]
Tusky opens links to toots and profile inside the app, but only when it can detect them as such. This detection is currently based on regexes (added in #568, extended in #1702 and #2304). However, the regexes are restrictive and don't admit e.g. PeerTube URLs, even though it's the third largest Fediverse software (by both user and node counts on the-federation.info). I bet they also exclude smaller, less popular software.

I suggest enhancing these checks with a domain check. There is a list of Fediverse nodes at nodes.fediverse.party, but you could obtain similar data from the-federation.info, fediverse.observer, and probably other places too. Even if the list is only updated at APK build time, it'd still be useful in detecting the smaller software makes with weird URL structures.

(Prompted by [this discussion](https://mastodon.online/@larsmb/109251890411023459)).

* * * *
- Tusky Version: 19
- Android Version: 10
- Android Device: OnePlus 5T
- Mastodon instance (if applicable):

- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.

**Comment by a1batross** [comment]
In case of PeerTube for example, opening video links in external app would be more useful. For example NewPipe has good support for PeerTube and even opening it in browser would allow peering.

It's not so hard to extend the list of regexes. I just found that it misses some cases. Ideally, an instance backend could detect if the link points to an ActivityPub object but oh well.

**Comment by Minoru** [comment]
@a1batross Good point, but one can always open in browser by tapping the three dots in the upper right corner. That's less convenient, but PeerTube is probably the only one that needs such treatment.

@connyduck Extending the list of regexes is a game of whack-a-mole. What's even more important, this game can't be won  because not all AP servers have a distinctive URL structure. For example, WordPress and WriteFreely both have pretty arbitrary URLs: https://loscimmiatoreseriale.altervista.org/tucandeo-esmee-bor-stotijn-northern-lights-matt-bukovski-remix-181 and https://write.tedomum.net/windy/le-cycle-de-leau Both of these projects are in top ten by number of nodes on the-federation.info (excluding Matrix and diaspora* which don't implement AP).

I agree that marking URLs at the instance backend would be the best solution, but detecting by domain is a stopgap that provides nearly the same accuracy and can be deployed today. Please consider it again.

**Comment by bt90** [comment]
Not sure about the URLs but it's rather simple to detect a fediverse domain using `/.well-known/nodeinfo`

https://video.blender.org/.well-known/nodeinfo
https://diaspora.freifunk-naila.net/.well-known/nodeinfo
https://mas.to/.well-known/nodeinfo
https://friendica.me/.well-known/nodeinfo

**Comment by Minoru** [comment]
@bt90 Fetching NodeInfo takes two HTTPS requests, which is likely to add noticeable latency. The original implementation in #568 applied search to every URL the user clicked on, and this was changed after a reviewer pointed out that it adds too much latency. That's where the regexes come in: *before* doing the search, Tusky checks if the URL even looks like it points to an ActivityPub object.

**Comment by bt90** [comment]
I don't see why we can't cache that information for a longer period of time. A domain is either part of the fediverse or it isn't.

**Comment by Minoru** [comment]
There are thousands of Fediverse instances, so there will be a lot of cache misses. That's why I'm suggesting to use a list of instances, which is essentially a pre-filled cache.
